src/D2L.CodeStyle.Analyzers/ApiUsage/RpcAnalyzer.IsValidParameterType.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage {
+
+	internal partial class RpcAnalyzer {
+
+		private static bool IsValidParameterType(
+			SyntaxNodeAnalysisContext context,
+			ITypeSymbol type,
+			RpcTypes rpcTypes
+		) {
+			if( rpcTypes.KnownRpcParameters.Contains( type ) ) {
+				return true;
+			}
+
+			if( type.TypeKind == TypeKind.Enum ) {
+				return true;
+			}
+
+			if( type.TypeKind == TypeKind.Array ) {
+				var arrayType = ( type as IArrayTypeSymbol ).ElementType as INamedTypeSymbol;
+				return IsValidParameterType( context, arrayType, rpcTypes );
+			}
+
+			if( !( type is INamedTypeSymbol namedType ) ) {
+				return false;
+			}
+
+			if( HasConstructorDeserializer( namedType, rpcTypes ) ) {
+				return true;
+			}
+
+			if( IsDeserializable( namedType, rpcTypes ) ) {
+				return true;
+			}
+
+			if( namedType.IsGenericType && SymbolEqualityComparer.Default.Equals( namedType.OriginalDefinition, rpcTypes.IDictionary ) ) {
+				var dictionaryTypes = namedType.TypeArguments;
+
+				if( !IsValidParameterType( context, dictionaryTypes[0], rpcTypes ) ) {
+					return false;
+				}
+
+				if( !IsValidParameterType( context, dictionaryTypes[1], rpcTypes ) ) {
+					return false;
+				}
+
+				return true;
+			}
+
+			return false;
+		}
+
+		private static bool HasConstructorDeserializer( INamedTypeSymbol type, RpcTypes rpcTypes ) {
+			if( rpcTypes.IDeserializer == null || rpcTypes.IDeserializer.Kind == SymbolKind.ErrorType ) {
+				return false;
+			}
+
+			foreach( var constructor in type.Constructors ) {
+				if( constructor.DeclaredAccessibility != Accessibility.Public ) {
+					continue;
+				}

```

Comment:
Force public, so it's not inconsistent between local usage and cross-assembly usage

Reply:
What do the reflecting BindingFlags look like in the LMS now? If they allow private constructors could we stop that too?